### PR TITLE
For #120: added multiple_while pattern

### PR DESCRIPTION
**Pull request by paulodamaso**
For #120: added multiple_while pattern and tests

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#135](https://github.com/yegor256/aibolit/issues/135) is assigned to @fanifieiev[/z](https://www.0crat.com/u/fanifieiev), here is [why](https://www.0crat.com/footprint/CT5UB7XAP/6b5f9b43-4e79-4f08-9634-14614038ebb9); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @acheshkov[/z](https://www.0crat.com/u/acheshkov) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: test/patterns/multiple_while/test_multiple_while.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+from aibolit.patterns.multiple_while.multiple_while import MultipleWhile
+from pathlib import Path
+
+
+class TestMultipleWhile(TestCase):
+    dir_path = Path(os.path.realpath(__file__)).parent
+    method_chain_finder = MultipleWhile()
+
+    def test_simple(self):
+        lines = self.method_chain_finder.value(Path(self.dir_path, 'Simple.java'))
```
Comment (fanifieiev) [inline test/patterns/multiple_while/test_multiple_while.py:34]: @paulodamaso It is better to:
`lines = MultipleWhile().value(Path(self.dir_path, 'Simple.java'))
`

File: test/patterns/multiple_while/test_multiple_while.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+import os
+from unittest import TestCase
+from aibolit.patterns.multiple_while.multiple_while import MultipleWhile
+from pathlib import Path
+
+
+class TestMultipleWhile(TestCase):
+    dir_path = Path(os.path.realpath(__file__)).parent
+    method_chain_finder = MultipleWhile()
```
Comment (fanifieiev) [inline test/patterns/multiple_while/test_multiple_while.py:31]: @paulodamaso Please remove this according to below.

File: aibolit/patterns/multiple_while/multiple_while.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+from collections import defaultdict
+import hashlib
+import itertools
+
+class MultipleWhile:
+
+    def __init__(self):
+        pass
+
+    def traverse_node(self, node, dict_with_chains, uuid_method):
```
Comment (fanifieiev) [inline aibolit/patterns/multiple_while/multiple_while.py:13]: @paulodamaso Where is this method used except inside itself as a recursion???

**Comment by paulodamaso** [comment]
@fanifieiev Changes made, please take a look

**Comment by paulodamaso** [comment]
@fanifieiev ping

File: aibolit/patterns/multiple_while/multiple_while.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import hashlib
+import itertools
+
+
+class MultipleWhile:
+
+    def __init__(self):
+        pass
+
+    def __file_to_ast(self, filename: str) -> javalang.ast.Node:
```
Comment (fanifieiev) [inline aibolit/patterns/multiple_while/multiple_while.py:13]: @paulodamaso Please merge with upstream and use `AST `class instead of this method

**Comment by 0crat** [comment]
@fanifieiev[/z](https://www.0crat.com/u/fanifieiev) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by acheshkov** [comment]
Is it ready to merge ?

**Comment by fanifieiev** [comment]
> Is it ready to merge ?

@acheshkov Not yet, waiting for @paulodamaso

**Comment by paulodamaso** [comment]
@fanifieiev Done, please take a look

**Comment by acheshkov** [comment]
@paulodamaso there are some conflicts, please take a look

**Comment by paulodamaso** [comment]
@acheshkov Done, please take a look

**Comment by acheshkov** [comment]
@rultor merge

**Comment by lyriccoder** [comment]
NOT ENOUGH tests!
1) CHECK while inside while
2) there is an anonymous class with while inside another while. 
3) Check `multiple try` tests, I see you copied a lot of code form that pattern
4) see issue when we have nested class https://github.com/yegor256/aibolit/issues/170

**Comment by fanifieiev** [comment]
> NOT ENOUGH tests!
> 
> 1. CHECK while inside while
> 2. there is an anonymous class with while inside another while.
> 3. Check `multiple try` tests, I see you copied a lot of code from that pattern
> 4. see issue when we have nested class #170
> 5. check lambda statements

@lyriccoder The original issue description is pretty simple and this PR covers that use case.
There might a number of other usecases(on top of what you suggest), so I think that we need to create another ticket and enhance the this pattern.
@paulodamaso pay attention.

**Comment by lyriccoder** [comment]
@fanifieiev That code was copied from my pattern (90% of code), I know where it could fail. I have fixed that code recently. Moreover, I am fixing all week such patterns where I found errors due to bad implementation like `pretty simple issue and this PR covers that use case`
I will reject all similar requests if you don't cover code tests. 1 test is not enough! We need working pattern! Otherwise it will affect the model

**Comment by acheshkov** [comment]
@rultor merge

**Comment by lyriccoder** [comment]
I've talked to @yegor256, he confirmed to merge it. I will write more tests by my own

**Comment by acheshkov** [comment]
@paulodamaso CI did not pass. Some problems with type checking. See the report above.

**Comment by paulodamaso** [comment]
@acheshkov I've reimplemented the method as per @lyriccoder comments, and fixed CI, please take a look

**Comment by acheshkov** [comment]
@rultor merge

**Comment by 0crat** [comment]
Code review was too long (8 days), architects (@acheshkov) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

**Pull request merged**